### PR TITLE
Add remote relations watcher

### PR DESCRIPTION
**Pull request by wallyworld**
## Description of change

Remove code that was added but no longer used in cross model relations.
Add a remote relations watcher for use in the firewaller worker.

## QA steps

bootstrap 
perform standard remote relation test
- add apps to different models
- relate
- test related app

### CODE REVIEW

File: apiserver/remoterelations/remoterelations.go [hunk 1]
```diff
@@ -618,3 +618,18 @@ func (api *RemoteRelationsAPI) WatchRemoteApplicationRelations(args params.Entit
 	}
 	return results, nil
 }
+
+// WatchRemoteApplications starts a strings watcher that notifies of the addition,
```
Comment (axw) [inline apiserver/remoterelations/remoterelations.go]: s/Applications/Relations/

File: apiserver/remoterelations/remoterelations.go [hunk 2]
```diff
@@ -618,3 +618,18 @@ func (api *RemoteRelationsAPI) WatchRemoteApplicationRelations(args params.Entit
...
+// removal, and lifecycle changes of remote applications in the model; and
```
Comment (axw) [inline apiserver/remoterelations/remoterelations.go]: s/applications/relations/
(and below)

File: state/watcher.go [hunk 3]
```diff
@@ -2733,3 +2724,51 @@ func (w *offeredApplicationsWatcher) loop() (err error) {
...
+			}
+			watchLogger.Errorf("could not load relation with key %v: %v", id, err)
+			return false
+		}
+
+		// Check to see if any of the endpoints are for remote applications.
+		remoteApps, closer := st.getCollection(remoteApplicationsC)
+		defer closer()
+
+		for _, ep := range doc.Endpoints {
```
Comment (axw) [inline state/watcher.go]: can you please make this a single query, using the `$in` operator?

File: state/watcher.go [hunk 4]
```diff
@@ -2733,3 +2724,51 @@ func (w *offeredApplicationsWatcher) loop() (err error) {
...
+		endpointFields := bson.D{{"endpoints", 1}}
+		var doc relationDoc
+
+		relations, closer := st.getCollection(relationsC)
+		defer closer()
+		if err := relations.FindId(id).Select(endpointFields).One(&doc); err != nil {
+			if err == mgo.ErrNotFound {
+				// If the relation can't be found, we allow it to be processed by the watcher.
+				// If it was a remote relation, it will trigger an event on transition to dead.
+				// If it wasn't a remote relation, it will result in a no-op as the relation was
```
Comment (axw) [inline state/watcher.go]: this feels a bit kludgey. can we keep track of which relations we've told the client about already (i.e. as a map/set closed over by this function)?

Comment (wallyworld) [inline state/watcher.go]: done

File: state/watcher.go (hunk 3 above)
Comment (wallyworld) [inline state/watcher.go]: done

File: apiserver/remoterelations/remoterelations.go (hunk 2 above)
Comment (wallyworld) [inline apiserver/remoterelations/remoterelations.go]: done

File: apiserver/remoterelations/remoterelations.go (hunk 1 above)
Comment (wallyworld) [inline apiserver/remoterelations/remoterelations.go]: done

**Comment by wallyworld** [comment]
I tried using a $lookup join but it didn't work with nested fields

**Comment by wallyworld** [comment]
$$merge$$

**Pull request merged**